commands/migratelxc.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package commands
+
+import (
+	"strings"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/lxc/lxd/shared/api"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/juju/1.25-upgrade/juju1/state"
+	"github.com/juju/1.25-upgrade/juju2/instance"
+)
+
+var migrateLXCDoc = ` 
+The purpose of the migrate-lxc command is to migrate all of the
+LXC containers in a 1.25 environment to LXD.
+`
+
+func newMigrateLXCCommand() cmd.Command {
+	command := &migrateLXCCommand{}
+	command.remoteCommand = "migrate-lxc-impl"
+	return wrap(command)
+}
+
+type migrateLXCCommand struct {
+	baseClientCommand
+	migrateDir string
+}
+
+func (c *migrateLXCCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "migrate-lxc",
+		Args:    "<environment name>",
+		Purpose: "migrate of all the LXC containers in the specified environment to LXD",
+		Doc:     migrateLXCDoc,
+	}
+}
+
+func (c *migrateLXCCommand) Init(args []string) error {
+	args, err := c.baseClientCommand.init(args)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return cmd.CheckEmpty(args)
+}
+
+var migrateLXCImplDoc = `
+
+migrate-lxc-impl must be executed on an API server machine of a 1.25
+environment.
+
+The command will migrate all LXC containers in the environment to LXD.
+
+`
+
+func newMigrateLXCImplCommand() cmd.Command {
+	return &migrateLXCImplCommand{}
+}
+
+type migrateLXCImplCommand struct {
+	baseRemoteCommand
+	containerName string
+}
+
+func (c *migrateLXCImplCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "migrate-lxc-impl",
+		Purpose: "controller aspect of migrate-lxc",
+		Doc:     migrateLXCImplDoc,
+	}
+}
+
+func (c *migrateLXCImplCommand) Run(ctx *cmd.Context) error {
+	st, err := c.getState(ctx)
+	if err != nil {
+		return errors.Annotate(err, "getting state")
+	}
+	defer st.Close()
+
+	// Collect LXC container machines by host.
+	byHost, err := getLXCContainers(st)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// List LXD containers on each host, so we know which LXC
```

Comment:
Maybe split this section out into a separate function that returns a `map[*state.Machine]map[string]*api.Container` (phew)? Like `getLXCContainers` above. Also it might be useful to differentiate in the function names that the LXC one is getting the info from state while the LXD one is getting it from the actual machines.